[Fix] If 'ninja' is not found, it installs it via pip

AllenFarcas opened the pull request:
# Description

Please include a brief summary of the changes, relevant motivation and context.

Fixes # (issue)

## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [X] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Infra/Build change
- [ ] Code refactoring

## Changes

Please list the changes introduced in this PR:

- If 'ninja' is not found, then 'ninja' will be installed via pup from setup.py
- Added checks with corresponding errors in build_ext.py and CMakeLists.txt

# Checklist:

- [x] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [ ] The functionality is complete
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

AllenFarcas committed:
[Fix] Ninja generator usage enforced explicitly, with checks added.

ipanfilo reviewed (commented):
I think it should also either explicitly use ninja generator when configure cmake extension in setup.py or in CMakeLists itself there should be check for generator

AllenFarcas commented:
@ipanfilo I just added in build_ext.py a runtime error if ninja is not found and also added that check in the CMakeLists a check for the generator that has a fatal error if ninja is not the CMAKE_GENERATOR.

Micky774 commented on build_tools/build_ext.py:
```diff
@@ -81,6 +81,8 @@ def _build_cmake(self, build_dir: Path, install_dir: Path) -> None:
         max_jobs = get_max_jobs_for_parallel_build()
         if found_ninja():
             configure_command.append("-GNinja")
+        else:
+            raise RuntimeError(f"This project requires the Ninja build system. Install it using 'pip install ninja'.")
```
```suggestion
        if not found_ninja():
            raise RuntimeError(f"This project requires the Ninja build system. Install it using 'pip install ninja'.")
        configure_command.append("-GNinja")
```

wangye805 commented on build_tools/build_ext.py:
```diff
@@ -79,8 +79,9 @@ def _build_cmake(self, build_dir: Path, install_dir: Path) -> None:
         max_jobs = get_max_jobs_for_parallel_build()
-        if found_ninja():
-            configure_command.append("-GNinja")
+        if not found_ninja():
```
Use rocm_build() to guard when we have different behavior with upstream NVTE

ipanfilo replied:
Throw error only on rocm build. I.e. add 
elif rocm_build():
   raise RauntimeError
to the original condition

ipanfilo commented on transformer_engine/common/CMakeLists.txt:
```diff
@@ -32,6 +32,10 @@ else()
   
 endif()
 
+if(NOT CMAKE_GENERATOR STREQUAL "Ninja")
```
Let's move it to ROCm specific section to let NV build still using Make

wangye805 replied:
Need to guard by USE_ROCM

wangye805 commented on setup.py:
```diff
@@ -117,6 +118,7 @@ def setup_requirements() -> Tuple[List[str], List[str], List[str]]:
     if not found_cmake():
         setup_reqs.append("cmake>=3.21")
     if not found_ninja():
+        subprocess.check_call([sys.executable, "-m", "pip", "install", "ninja"])
```
I don't think we need to call pip install ninja directly. 

TE installation should be able to install required packages/dependencies once we put it in setup_reqs

AllenFarcas replied:
The setup_reqs actually contains ninja if it is not installed and yet, it does not install it. I searched more info on this and found that setup_requires is considered a legacy mechanism. PEP 518 introduced the pyproject.toml file and a mechanism to specify your build-time dependencies. 

Do you think that leaving this install of ninja here is ok @wangye805 @ipanfilo @wenchenvincent ?

https://stackoverflow.com/questions/73257839/setup-py-install-is-deprecated-warning-shows-up-every-time-i-open-a-terminal-i

https://blog.ganssle.io/articles/2021/10/setup-py-deprecated.html#summary

wangye805 replied:
Was this feature already deprecated? In other words, was the root cause of setup_reqs contains ninja but not installed due to that this feature deprecated? 

If so, we can temporarily keep this "pip install ninja". 

Just notice NV upstream already switched to pyproject.toml: https://github.com/NVIDIA/TransformerEngine/pull/1852/files

AllenFarcas replied:
This feature was already deprecated. It appears for quite some time, setup.py wasn't able to reliably install and ensure the "setup_reqs" are met. Due to this unreliability, we currently can patch with the "pip install ninja".

ipanfilo replied:
Indeed setup_reqs does not work. So I think install it with pip here is OK. One possible thing to improve is make ninja>=1.10 as it is minimal version for aotriton building but I think pip install will actually have something newer anyways

AllenFarcas committed:
Added USE_ROCM guard in CMakeLists and rocm_buil() guard in build_ext.py
